### DISCUSSION TITLE
How to boot from bank 1 or 2 / Dual Bank Flash / STM32L082

### DISCUSSION THREAD

**Discussion by JeroenIoT** [stm32duino/Arduino_Core_STM32#1866] [Q&A] [answered]
Hello,

I am trying to create code which is able to run from flash bank1 or from flash bank2 (using the dual bank STM32L0 architecture)

For this I am using platformio (stm32duino):
* In platformio.ini I created 2 environments (bank 1 and bank 2)

```
[env:STM32L082_bank1]
; no special options

[env:STM32L082_bank2]
extra_scripts = set_bank2.py
build_flags = -Wl,--defsym=LD_MAX_DATA_SIZE=0x5000 -Wl,--defsym=LD_MAX_SIZE=0x30000 -Wl,--defsym=LD_FLASH_OFFSET=0x18000
```

* For bank2 "set_bank2.py" script in order to change VECT_TAB_OFFSET:
```
Import("env")
offset = 0x18000 # offset from 0x8000000
env.Append(
    CPPDEFINES=[("VECT_TAB_OFFSET", "%s" % hex(offset))],
)
# remove old 0-offset, inject new one
linkflags = env["LINKFLAGS"]
linkflags = [x for x in linkflags if not str(x).startswith("-Wl,--defsym=LD_FLASH_OFFSET=")]
linkflags.append("-Wl,--defsym=LD_FLASH_OFFSET=%s" % hex(offset))
env["LINKFLAGS"] = linkflags 
```

* The generated BIN files are uploaded (bank1 to to 0x0800 0000 and bank2 to 0x0801 8000). 
* Bank1 is running fine 
* To run the code on bank2 I alter the OB BFB2 (using CUBEProgrammer)

However I do not get bank2 code running. Does anybody have suggestions or a code example for platformio (stm32duino based).

**Comment by fpistm** [comment]
Hi,
I never used DUAL bank. And don't know PIO.
you will probably have better support on PIO or stm32duino.com forums.
Maybe this AP can help: https://www.st.com/resource/en/application_note/an4767-onthefly-firmware-update-for-dual-bank-stm32-microcontrollers-stmicroelectronics.pdf

**Comment by ABOSTM** [comment]
@JeroenIoT,
Looking at Reference Manual:
RM0376 Rev 7 ,  paragraph 3.3.2 Dual-bank boot capability

> • When the BFB2 bit is set and the boot pins are configured to boot from Flash memory
> (BOOT0 = 0 and BOOT1 = x), the device maps the System memory at address 0. It
> boots from the System memory after reset and Standby and executes (during
> approximately 440 μs) the embedded bootloader code which implements the dualbank
> boot mechanism:
> a) The System memory code first checks Bank 2. If it contains a valid code (see note
> below), **it sets the UFB bit** in SYSCFG_CFGR1 register **to map Bank 2 at address
> 0x0800 0000**, ...
> ...

Thus you binary for bank2 must be compiled and linked as if it is executed @0x08000000, so without offset 0x18000

**Reply by JeroenIoT** [reply]
Hi Alexandre,

Thanks for the feedback.  Using VECT_TAB_OFFSET=0x18000 worked in the end for me, however it makes it a bit more complex as I have 2 different binary files (one for each bank).

So I started to test your suggestion. Unfortunately I could not get it working (yet) - below my results:

* Assumed BANK2 binary is identical to BANK1 binary 
* Set `LD_FLASH_OFFSET` to 0x18000 (flash of BANK2 binary starts at 0x08018000)
* Checked banks after flashing: start (msp and pc) of bank1 and bank2 looks good
* However when swapping to BANK2 the code does not run (tried swap bank via app code and via CUBE Programmer - both had the same result).
* Below elf file of bank2 (which is identical to bank1 elf)
```
.pio\build\STM32L082_bank2\firmware.elf:     file format elf32-littlearm

Disassembly of section .isr_vector:

08000000 <g_pfnVectors>:
 8000000:       20005000        andcs   r5, r0, r0
 8000004:       080031f1        stmdaeq r0, {r0, r4, r5, r6, r7, r8, ip, sp}

```

Suggestions are appreciated :-)

**Comment by ABOSTM** [comment]
Hi @JeroenIoT ,
I made the following test which is working fine (which confirm the description from my previous post):
* I build a binary for bank1 (a blinky example)
* I program bank1 with this binary (using cube programmer), and it work fine
* I erase the whole flash (bank1 and bank2)
* without recompilation I program flash bank2 with the exact same binary (not the elf) using cube programmer
 This is the only step where I used a 0x18000 offset
 Of course it didn't work yet but ...
* .. I set BFB2 to 1 (cubeProgrammer) and then blinks worked again

I don't know about platformio.
I wonder whether it is a good thing in your setup to have LD_FLASH_OFFSET = 0x18000
as it impacts linker script (and thus linker stage),
For test purpose you can try to keep LD_FLASH_OFFSET =0 and to flash with cubeprogrammer.
Note: with Cube programmer bank2 is presented @0x18000 whatever BFB2 value

**Comment by JeroenIoT** [comment]
Hi Alexandre,

That is really interesting - I did the following

* Basic code to serialprint couple of lines
* LD_FLASH_OFFSET =0x0 
* Generate BIN file
* Upload via CubeProgrammer to 0x08000000 (BFB2 = 0 )
* Serial prints are shown in serial monitor - ALL GOOD 
* Now erase full flash
* Upload via CubeProgrammer to 0x08018000 and set BFB2 = 1 (see below)
* ... and there it ends. The code does not start to run from bank2

![image](https://user-images.githubusercontent.com/14053463/202235703-4de203f5-ed96-4820-907f-e2090169c062.png)

Do anybody have a sample bin with serial output? Would like to test it and understand why mine is not working with single BIN for both banks

**Reply by JeroenIoT** [reply]
I just found out that when I change the SystemInit(void) I can get it starting from bank2 (from `system_stm32l0xx.c`)

```
  /* Configure the Vector Table location add offset address ------------------*/
#if defined (USER_VECT_TAB_ADDRESS)
  //SCB->VTOR = VECT_TAB_BASE_ADDRESS | VECT_TAB_OFFSET; /* Vector Table Relocation in Internal SRAM */
  //SCB->VTOR = 0x08018000UL;
#endif /* USER_VECT_TAB_ADDRESS */
  SCB->VTOR = 0x08018000UL;
  //SCB->VTOR = 0x08000000UL;
}
```

**Comment by ABOSTM** [comment]
I just remember,
there maybe a hardware bug depending on your hardware revision:
https://www.stm32duino.com/viewtopic.php?p=9512&sid=c7afedbe3211db4b141e89db1586af30#p9512

Can you check you hardware revision ?

**Reply by JeroenIoT** [reply]
I think my hardware version is revision Z.

Photo below and reading out register I get following information

* devID               : 447
* revID               : 2008

![image](https://user-images.githubusercontent.com/14053463/202534598-a0c18794-12d8-4a90-a9fd-6b6894d381cf.png)

When reading the errata document I do not think it is my hardware version (afai can see) 

Following from [ST doc:](https://www.st.com/resource/en/errata_sheet/es0292-stm32l07xxxl08xxx-device-errata-stmicroelectronics.pdf) 
```
The following limitations are fixed in device revision Z:
• BFB2 bit does not select boot from Bank 2 system memory

```

**Comment by JeroenIoT** [comment]
Quick update

* I removed offsets and created 1 environment. 
* When I add the `SCB->VTOR = 0x08000000UL;` to the end of the `void SystemInit (void)` the dual bank works
* In this case I generate a single BIN, flash it in both BANKs and it works good

So where is the SCB->VTOR normally set? In the original SystemInit(void) file it is not set and it works when using bank1 only. 

@ABOSTM - how does the SystemInit() looks like for the stm32 processor you used in the test above? Was SCB->VTOR set?

**Comment by ABOSTM** [comment, marked as answer]
Yes it is done in at the end of `void SystemInit (void)`
but it is already done in latest version:
https://github.com/stm32duino/Arduino_Core_STM32/blob/75b051e50cb2b02cb110d69433917e5c6d14ad96/system/STM32L0xx/system_stm32l0xx.c#L174

So maybe you are not on latest version.
You you check with version 2.3.0 of Arduino_Core_STM32, or later ?

**Reply by JeroenIoT** [reply]
Great - that means my problem was tackled.:-)

* Yes I am using öld version - PIO is only at version STM32duino v2.2.0
* In this old v2.2.0 version the #if statements were not good when activating the dual bank.
* Tested the new conditions and it all works now
* So using the 2.3.0 version or alternatively one can add to the platformio.ini the following: `build_flags = -D VECT_TAB_OFFSET=0x00000000U`

Thanks a lot!
